src/crypto/secp156k1_provider/secp256k1_provider_impl.cpp

```diff
@@ -0,0 +1,132 @@
+      return KeyGeneratorError::KEY_GENERATION_FAILED;
+    }
+    return KeyPair{privateKey, publicKey};
+  }
+
+  outcome::result<PublicKey> Secp256k1ProviderImpl::derivePublicKey(
+      const PrivateKey &privateKey) const {
+    OUTCOME_TRY(key, bytesToPrivateKey(privateKey));
+    PublicKey publicKey{};
+    uint8_t *publicKeyPtr = publicKey.data();
+    ssize_t generatedLength = i2o_ECPublicKey(key.get(), nullptr);
+    if (generatedLength != publicKey.size()) {
+      return KeyGeneratorError::KEY_GENERATION_FAILED;
+    }
+    i2o_ECPublicKey(key.get(), &publicKeyPtr);
+    return publicKey;
+  }
+
+  outcome::result<Signature> Secp256k1ProviderImpl::sign(
+      gsl::span<uint8_t> message, const PrivateKey &privateKey) const {
+    common::Hash256 digest = sha256(message);
+    OUTCOME_TRY(key, bytesToPrivateKey(privateKey));
+    std::shared_ptr<ECDSA_SIG> ecSignature{
+        ECDSA_do_sign(digest.data(), digest.size(), key.get()), ECDSA_SIG_free};
+    if (ecSignature == nullptr) {
+      return CryptoProviderError::SIGNATURE_GENERATION_FAILED;
+    }
+    Signature signature;
+    ssize_t signatureLength = i2d_ECDSA_SIG(ecSignature.get(), nullptr);
+    if (signatureLength < 0) {
+      return CryptoProviderError::SIGNATURE_GENERATION_FAILED;
+    }
+    signature.resize(signatureLength);
+    uint8_t *signaturePtr = signature.data();
+    i2d_ECDSA_SIG(ecSignature.get(), &signaturePtr);
+    return signature;
+  }
+
+  outcome::result<bool> Secp256k1ProviderImpl::verify(
+      gsl::span<uint8_t> message, const Signature &signature,
+      const PublicKey &publicKey) const {
+    common::Hash256 digest = sha256(message);
+    OUTCOME_TRY(key, bytesToPublicKey(publicKey));
+    int result = ECDSA_verify(0, digest.data(), digest.size(), signature.data(),
+                              signature.size(), key.get());
+    if (result < 0) {
+      return CryptoProviderError::SIGNATURE_VERIFICATION_FAILED;
+    }
+    return result == 1;
+  }
+
+  outcome::result<std::shared_ptr<EC_KEY>>
+  Secp256k1ProviderImpl::bytesToPrivateKey(const PrivateKey &input) {
+    std::shared_ptr<EC_KEY> key{EC_KEY_new_by_curve_name(NID_secp256k1),
+                                EC_KEY_free};
+    if (key.get() == nullptr) {
+      return KeyGeneratorError::INTERNAL_ERROR;
+    }
+    std::unique_ptr<BIGNUM, void (*)(BIGNUM *)> keyNum{
+        BN_bin2bn(input.data(), input.size(), nullptr), BN_free};
+    if (keyNum == nullptr) {
+      return KeyGeneratorError::INTERNAL_ERROR;
+    }
+    EC_KEY_set_private_key(key.get(), keyNum.get());
+    std::unique_ptr<EC_POINT, void (*)(EC_POINT *)> point{
+        EC_POINT_new(EC_KEY_get0_group(key.get())), EC_POINT_free};
+    if (point == nullptr) {
+      return KeyGeneratorError::INTERNAL_ERROR;
+    }
+    EC_KEY_set_conv_form(key.get(), POINT_CONVERSION_COMPRESSED);
+    EC_POINT_mul(EC_KEY_get0_group(key.get()), point.get(), keyNum.get(),
```

Comment:
check return code
[The following functions return 1 on success or 0 on error: EC_POINT_add, EC_POINT_dbl, EC_POINT_invert, EC_POINT_make_affine, EC_POINTs_make_affine, EC_POINTs_make_affine, EC_POINT_mul, EC_POINTs_mul and EC_GROUP_precompute_mult.](https://www.openssl.org/docs/man1.0.2/man3/EC_POINT_mul.html)